File: src/tests/Behavioral/ReflectFieldMetadata/ReflectFieldMetadata.go
// Struct-field METADATA through reflect, three roots measured by reflect's own suite and pinned here
// against `go run` (increment E2 of the reflect tail):
//
//   1. Go's flag.ro(): an EMBEDDED read-only field's elements stay read-only through Index and
//      Slice, so CanSet answers false two hops down (TestIssue22031).
//   2. FieldByName's MULTIPLICITY rule: an embedded type reached twice at one depth annihilates
//      itself -- its field is absent, not found once (TestFieldByName's S3.B / S10.X / S14.X).
//   3. StructField.PkgPath is the package that DECLARED the field, even when the struct is a defined
//      type over another package's struct (TestFieldPkgPath's localOtherPkgFields row).
//
//   4. StructField.Anonymous for an embedded BUILTIN (`struct{ int }`, `struct{ *int }`): a plain field
//      the converter stamps [GoEmbedded] (increment E2b -- the other half of TestFieldPkgPath), beside
//      the controls that must NOT read embedded (a plain field of a named-int type) and the embedded
//      NAMED int the struct-wrapper route already reported.
//
// A second shape withheld for a different reason: a struct EMBEDDING `twice` (`type deeper struct{ twice }`,
// the clause that an annihilation at one depth inhibits every deeper match) does not compile through
// go2cs-gen today. TypeGenerator's transitive promotion walk (StructTypeTemplate.getStructMembers) guards
// against cycles with ONE `seenTypes` set shared across sibling branches, so `base` -- reached through
// `viaX` AND `*viaY` at the same depth -- is walked once and its `B` counted once: `deeper` promotes
// `B => ref twice.B` while `twice`'s own shell (each embed walked afresh) correctly promotes no `B`, and
// the forwarder binds a nonexistent hop (CS1061 x2 in go.main_package.deeper.g.cs). It is the same
// visited-vs-multiplicity error E2 fixes in reflect's promotedFieldByName, on the generator's side; the
// shape returns here when the walk's guard is scoped to the current PATH (its own increment, route #7 gates).
package main

import (
	"fmt"
	"reflect"

	"ReflectFieldMetadata/fieldlib"
)

// --- root 1: flag.ro() ---
type sElem []struct{ C int }
type embeds struct{ sElem }   // the slice type is EMBEDDED and unexported: flagEmbedRO
type holds struct{ f sElem }  // a plain unexported field: flagStickyRO

// --- root 2: multiplicity ---
type base struct{ B int }
type viaX struct{ base }
type viaY struct{ base }
type twice struct { // base is reached through viaX AND *viaY at depth 2: B annihilates
	viaX
	*viaY
	D int
}
type once struct { // base reached ONCE: B is found at depth 2
	viaX
	D int
}

// --- root 3: PkgPath through a defined type over a foreign struct ---
type local fieldlib.Outer

// --- root 4 (E2b): Anonymous for an embedded builtin ---
type myInt int
type embedsInt struct{ int }        // embedded PREDECLARED type: Anonymous, named "int"
type embedsIntPtr struct{ *int }    // embedded POINTER to a predeclared type: Anonymous, named "int"
type holdsNamed struct{ n myInt }   // a plain field of a named-int type: NOT embedded (control)
type embedsNamed struct{ myInt }    // an embedded NAMED int: Anonymous through the struct-wrapper route (control)

func main() {
	// 1. CanSet through an embedded read-only slice's element's field.
	fmt.Println("CanSet via embedded  :", reflect.ValueOf(embeds{sElem{{}}}).Field(0).Index(0).Field(0).CanSet())
	fmt.Println("CanSet via unexported:", reflect.ValueOf(holds{sElem{{}}}).Field(0).Index(0).Field(0).CanSet())
	// and the same read-only survives a Slice window
	fmt.Println("CanSet via Slice     :", reflect.ValueOf(embeds{sElem{{}}}).Field(0).Slice(0, 1).Index(0).Field(0).CanSet())

	// 2. FieldByName: found / not found, and the index path when found.
	_, foundTwice := reflect.TypeOf(twice{}).FieldByName("B")
	fOnce, foundOnce := reflect.TypeOf(once{}).FieldByName("B")
	fD, foundD := reflect.TypeOf(twice{}).FieldByName("D")
	fmt.Println("twice.B  found:", foundTwice)
	fmt.Println("once.B   found:", foundOnce, "index:", fOnce.Index)
	fmt.Println("twice.D  found:", foundD, "index:", fD.Index)
	// the Value side agrees with the Type side
	fmt.Println("twice.B  value valid:", reflect.ValueOf(twice{}).FieldByName("B").IsValid())

	// 3. PkgPath of a field declared in another package, reached through a defined local type. The
	// rows compare PATHS rather than print them, so the guard measures E2's root -- the declaring
	// package survives the defined-type hop -- and not the spelling of a sub-library's import path.
	lt := reflect.TypeOf(local{})
	ot := reflect.TypeOf(fieldlib.Outer{})
	mine := reflect.TypeOf(struct{ u int }{}).Field(0).PkgPath
	fmt.Println("local.Field(0) exported, PkgPath empty:", lt.Field(0).IsExported(), lt.Field(0).PkgPath == "")
	fmt.Println("local.Field(1) exported:", lt.Field(1).IsExported())
	fmt.Println("local.Field(1) PkgPath == Outer.Field(1) PkgPath:", lt.Field(1).PkgPath == ot.Field(1).PkgPath)
	fmt.Println("local.Field(1) PkgPath is foreign (not this package, not empty):", lt.Field(1).PkgPath != mine, lt.Field(1).PkgPath != "")
	fmt.Println("anon unexported PkgPath is this package:", mine == "main")

	// 4. Anonymous for an embedded builtin, with its controls.
	anon := func(label string, t reflect.Type) {
		f := t.Field(0)
		fmt.Printf("%-14s name=%-5s anonymous=%-5v type=%-6v pkgPath=%q\n", label, f.Name, f.Anonymous, f.Type, f.PkgPath)
	}
	anon("struct{ int }", reflect.TypeOf(embedsInt{}))
	anon("struct{ *int }", reflect.TypeOf(embedsIntPtr{}))
	anon("struct{ n T }", reflect.TypeOf(holdsNamed{}))
	anon("struct{ myInt }", reflect.TypeOf(embedsNamed{}))
}

File: src/tests/Behavioral/ReflectFieldMetadata/ReflectFieldMetadata.cs
namespace go;

using fmt = fmt_package;
using reflect = reflect_package;
using fieldlib = ReflectFieldMetadata.fieldlib_package;
using ReflectFieldMetadata;

partial class main_package {

[GoType("dyn")] partial struct sElemᴛ1 {
    public nint C;
}

[GoType("[]sElemᴛ1")] partial struct sElem;

[GoType] partial struct embeds {
    internal partial ref sElem sElem { get; }
}

[GoType] partial struct holds {
    internal sElem f;
}

[GoType] partial struct @base {
    public nint B;
}

internal static nint M(this @base b) {
    return b.B;
}

[GoType] partial struct viaX {
    internal partial ref @base @base { get; }
}

[GoType] partial struct viaY {
    internal partial ref @base @base { get; }
}

[GoType] partial struct twice {
    internal partial ref viaX viaX { get; }
    internal partial ref ж<viaY> viaY { get; }
    public nint D;
}

[GoType] partial struct once {
    internal partial ref viaX viaX { get; }
    public nint D;
}

[GoType] partial struct deeper {
    internal partial ref twice twice { get; }
}

[GoType("ReflectFieldMetadata.fieldlib_package.Outer")] partial struct local;

[GoType("num:nint")] partial struct myInt;

[GoType] partial struct embedsInt {
    [GoEmbedded] internal nint @int;
}

[GoType] partial struct embedsIntPtr {
    [GoEmbedded] internal ж<nint> @int;
}

[GoType] partial struct holdsNamed {
    internal myInt n;
}

[GoType] partial struct embedsNamed {
    internal partial ref myInt myInt { get; }
}

// Hoisted @string literals (single allocation; Go keeps these in RODATA)
private static readonly object canSetViaEmbeddedˢ = (@string)"CanSet via embedded  :"u8;
private static readonly object canSetViaUnexportedˢ = (@string)"CanSet via unexported:"u8;
private static readonly object canSetViaSliceˢ = (@string)"CanSet via Slice     :"u8;
private static readonly object twiceBFoundˢ = (@string)"twice.B  found:"u8;
private static readonly object onceBFoundˢ = (@string)"once.B   found:"u8;
private static readonly object indexˢ = (@string)"index:"u8;
private static readonly object twiceDFoundˢ = (@string)"twice.D  found:"u8;
private static readonly object twiceBValueValidˢ = (@string)"twice.B  value valid:"u8;
private static readonly object deeperBFoundˢ = (@string)"deeper.B found:"u8;
private static readonly object valueValidˢ = (@string)" value valid:"u8;
private static readonly object deeperDFoundˢ = (@string)"deeper.D found:"u8;
private static readonly object valueˢ = (@string)" value:"u8;
private static readonly object mPromotedTwiceˢ = (@string)"M promoted -- twice:"u8;
private static readonly object deeperˢ = (@string)" deeper:"u8;
private static readonly object onceˢ = (@string)" once:"u8;
private static readonly object localField0Exportedˢ = (@string)"local.Field(0) exported, PkgPath empty:"u8;
private static readonly object localField1Exportedˢ = (@string)"local.Field(1) exported:"u8;
private static readonly object localField1PkgPathOuterˢ = (@string)"local.Field(1) PkgPath == Outer.Field(1) PkgPath:"u8;
private static readonly object localField1PkgPathIsˢ = (@string)"local.Field(1) PkgPath is foreign (not this package, not empty):"u8;
private static readonly object anonUnexportedPkgPathIsˢ = (@string)"anon unexported PkgPath is this package:"u8;
private static readonly @string structIntˢ = "struct{ int }"u8;
private static readonly @string structIntˢ2 = "struct{ *int }"u8;
private static readonly @string structNTˢ = "struct{ n T }"u8;
private static readonly @string structMyIntˢ = "struct{ myInt }"u8;

[GoType("dyn")] internal partial struct main_i {
    internal nint u;
}

internal static void Main() {
    fmt.Println(canSetViaEmbeddedˢ, reflect.ValueOf(new embeds(new sElem(new sElemᴛ1[]{new()}.slice()))).Field(0).Index(0).Field(0).CanSet());
    fmt.Println(canSetViaUnexportedˢ, reflect.ValueOf(new holds(new sElem(new sElemᴛ1[]{new()}.slice()))).Field(0).Index(0).Field(0).CanSet());
    fmt.Println(canSetViaSliceˢ, reflect.ValueOf(new embeds(new sElem(new sElemᴛ1[]{new()}.slice()))).Field(0).Slice(0, 1).Index(0).Field(0).CanSet());
    var (_, foundTwice) = reflect.TypeOf(new twice(nil)).FieldByName("B"u8);
    var (fOnce, foundOnce) = reflect.TypeOf(new once(nil)).FieldByName("B"u8);
    var (fD, foundD) = reflect.TypeOf(new twice(nil)).FieldByName("D"u8);
    fmt.Println(twiceBFoundˢ, foundTwice);
    fmt.Println(onceBFoundˢ, foundOnce, indexˢ, fOnce.Index);
    fmt.Println(twiceDFoundˢ, foundD, indexˢ, fD.Index);
    fmt.Println(twiceBValueValidˢ, reflect.ValueOf(new twice(nil)).FieldByName("B"u8).IsValid());
    var (_, foundDeeperB) = reflect.TypeOf(new deeper(nil)).FieldByName("B"u8);
    var (fDeeperD, foundDeeperD) = reflect.TypeOf(new deeper(nil)).FieldByName("D"u8);
    fmt.Println(deeperBFoundˢ, foundDeeperB, valueValidˢ, reflect.ValueOf(new deeper(nil)).FieldByName("B"u8).IsValid());
    fmt.Println(deeperDFoundˢ, foundDeeperD, indexˢ, fDeeperD.Index, valueˢ, reflect.ValueOf(new deeper(nil)).FieldByName("D"u8).IsValid());
    var (_, mTwice) = reflect.TypeOf(new twice(nil)).MethodByName("M"u8);
    var (_, mDeeper) = reflect.TypeOf(new deeper(nil)).MethodByName("M"u8);
    var (_, mOnce) = reflect.TypeOf(new once(nil)).MethodByName("M"u8);
    fmt.Println(mPromotedTwiceˢ, mTwice, deeperˢ, mDeeper, onceˢ, mOnce);
    var lt = reflect.TypeOf(new local(new fieldlib.Outer(nil)));
    var ot = reflect.TypeOf(new fieldlib.Outer(nil));
    @string mine = reflect.TypeOf(new main_i()).Field(0).PkgPath;
    fmt.Println(localField0Exportedˢ, lt.Field(0).IsExported(), lt.Field(0).PkgPath == ""u8);
    fmt.Println(localField1Exportedˢ, lt.Field(1).IsExported());
    fmt.Println(localField1PkgPathOuterˢ, lt.Field(1).PkgPath == ot.Field(1).PkgPath);
    fmt.Println(localField1PkgPathIsˢ, lt.Field(1).PkgPath != mine, lt.Field(1).PkgPath != ""u8);
    fmt.Println(anonUnexportedPkgPathIsˢ, mine == "main"u8);
    void anon(@string label, reflectꓸType t) {
        var f = t.Field(0);
        fmt.Printf("%-14s name=%-5s anonymous=%-5v type=%-6v pkgPath=%q\n"u8, label, f.Name, f.Anonymous, f.Type, f.PkgPath);
    }
    anon(structIntˢ, reflect.TypeOf(new embedsInt(nil)));
    anon(structIntˢ2, reflect.TypeOf(new embedsIntPtr(nil)));
    anon(structNTˢ, reflect.TypeOf(new holdsNamed(nil)));
    anon(structMyIntˢ, reflect.TypeOf(new embedsNamed(nil)));
}

} // end main_package
